### PR TITLE
git.py: Fail on different branches for same repo

### PR DESCRIPTION
**Pull request by Mishytca**
```
  doma:
    sources:
      - type: git url: https://github.com/xen-troops/meta-example.git rev: "ABCDEF"
    ...
  domu:
    sources:
      - type: git url: https://github.com/xen-troops/meta-example.git rev: "branch_A"
```

For doma and domu, the same GitHub repository is specified, but with different branches.

In the case of such an erroneous YAML file, Moulin will not highlight this error. Consequently, the user will also spend significant effort to understand the reason for the incorrect functioning.

This patch adds an additional check to the git.py file. Now, if the repository branches don't match, an error is displayed in the console, and the generation of the build.ninja file is halted.

### CODE REVIEW

**Review by lorc — CHANGES_REQUESTED** [review]
I have a slight sense of deja vu...

It would be really nice to attribute the original author of the code. There is the `Co-Authored-By:` tag exactly for this purpose.

File: moulin/fetchers/git.py [hunk 1]
```diff
@@ -63,11 +63,17 @@ def gen_fetch(self):
         clone_target = self.git_dir
         checkout_stamp = create_stamp_name(self.build_dir, self.url, "checkout")
 
-        # Do not checkout repos for the second time
-        if checkout_stamp in _SEEN_REPOS:
-            return checkout_stamp
+        if checkout_stamp in _SEEN_REPOS_REV:
+            if self.git_rev != _SEEN_REPOS_REV[checkout_stamp]:
+                # Do not checkout repos for the second time
+                raise Exception(f"ERROR: Repository {self.url} has two revisions '{self.git_rev}' "
```
[the remark this answers is not in the document]
Comment (Mishytca) [reply moulin/fetchers/git.py]: Hi @lorc , done.

File: moulin/fetchers/git.py [hunk 2]
```diff
@@ -63,11 +63,17 @@ def gen_fetch(self):
...
+                                "and '{_SEEN_REPOS_REV[checkout_stamp]}'")
+            else:
+                # Fail on occurrence of different revision for the already downloaded repository
```
[the remark this answers is not in the document]
Comment (Mishytca) [reply moulin/fetchers/git.py:69]: Hi @lorc , done.

**Review by lorc — APPROVED** [review]
`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

**Pull request merged**